plugins/hls-eval-plugin/src/Ide/Plugin/Eval/Types.hs

```diff
@@ -30,17 +33,75 @@ module Ide.Plugin.Eval.Types
+import           Control.Lens
+import           Data.Aeson                      (FromJSON, ToJSON)
+import           Data.List                       (partition)
+import           Data.List.NonEmpty              (NonEmpty)
+import           Data.Map.Strict                 (Map)
+import           Data.String                     (IsString (..))
+import qualified Data.Text                       as T
+import           Development.IDE                 (Range, RuleResult)
+import qualified Development.IDE.Core.Shake      as Shake
+import qualified Development.IDE.GHC.Compat.Core as Core
 import           Development.IDE.Graph.Classes
-import           GHC.Generics                  (Generic)
-import           Language.LSP.Protocol.Types   (TextDocumentIdentifier)
-import qualified Text.Megaparsec               as P
+import           GHC.Generics                    (Generic)
+import           Ide.Logger
+import           Ide.Plugin.Eval.GHC             (showDynFlags)
+import           Ide.Plugin.Eval.Util
+import           Language.LSP.Protocol.Types     (TextDocumentIdentifier,
+                                                  TextEdit)
+import qualified System.Time.Extra               as Extra
+import qualified Text.Megaparsec                 as P
+
+data Log
+    = LogShake Shake.Log
+    | LogCodeLensFp FilePath
+    | LogCodeLensComments Comments
+    | LogExecutionTime T.Text Extra.Seconds
+    | LogTests !Int !Int !Int !Int
+    | LogRunTestResults [T.Text]
+    | LogRunTestEdits TextEdit
+    | LogEvalFlags [String]
+    | LogEvalPreSetDynFlags Core.DynFlags
+    | LogEvalParsedFlags
+        (Either
+            Core.GhcException
+            (Core.DynFlags, [Core.Located String], [Core.Warn]))
+    | LogEvalPostSetDynFlags Core.DynFlags
+    | LogEvalStmtStart String
+    | LogEvalStmtResult (Maybe [T.Text])
+    | LogEvalImport String
+    | LogEvalDeclaration String
+
+instance Pretty Log where
+    pretty = \case
+        LogShake shakeLog -> pretty shakeLog
+        LogCodeLensFp fp -> "fp" <+> pretty fp
+        LogCodeLensComments comments -> "comments" <+> viaShow comments
+        LogExecutionTime lbl duration -> pretty lbl <> ":" <+> pretty (Extra.showDuration duration)
+        LogTests nTests nNonSetupSections nSetupSections nLenses -> "Tests" <+> fillSep
+            [ pretty nTests
+            , "tests in"
+            , pretty nNonSetupSections
+            , "sections"
+            , pretty nSetupSections
+            , "setups"
+            , pretty nLenses
+            , "lenses."
+            ]
+        LogRunTestResults results ->  "TEST RESULTS" <+> viaShow results
+        LogRunTestEdits edits -> "TEST EDIT" <+> viaShow edits
+        LogEvalFlags flags -> "{:SET" <+> pretty flags
+        LogEvalPreSetDynFlags dynFlags -> "pre set" <+> pretty (showDynFlags dynFlags)
+        LogEvalParsedFlags eans -> "parsed flags" <+> viaShow (eans
+              <&> (_1 %~ showDynFlags >>> _3 %~ prettyWarnings))
```

Comment:
plugins/hls-eval-plugin/src/Ide/Plugin/Eval/Types.hs:99:43: error: [GHC-18872]
Error:     • Couldn't match type: [Core.Warn]
                     with: GHC.Types.Error.Messages
                             GHC.Driver.Errors.Types.DriverMessage
        arising from a functional dependency between:
          constraint ‘Field3
                        ([Char], [Core.Located String], [Core.Warn])
                        ([Char], [Core.Located String], [Char])
                        (GHC.Types.Error.Messages GHC.Driver.Errors.Types.DriverMessage)
                        String’
            arising from a use of ‘_3’
          instance ‘Field3 (a, b, c) (a, b, c') c c'’ at <no location info>
    • In the first argument of ‘(%~)’, namely ‘_3’
      In the second argument of ‘(>>>)’, namely ‘_3 %~ prettyWarnings’
      In the second argument of ‘(<&>)’, namely
        ‘(_1 %~ showDynFlags >>> _3 %~ prettyWarnings)’
   |
99 |               <&> (_1 %~ showDynFlags >>> _3 %~ prettyWarnings))
   |                                           ^^

Reply:
Thanks, GHC API compatibility remains a headache :)